core/src/Streamly/Internal/Data/Fold/Window.hs

```diff
@@ -351,3 +360,82 @@ mean = Fold.teeWith (/) sum length
 {-# INLINE range #-}
 range :: (Monad m, Num a, Ord a) => Fold m (a, Maybe a) a
 range = Fold.teeWith (-) maximum minimum
+
+{-# INLINE sumWith #-}
+sumWith :: forall m a. (Monad m, Num a)
+    => Fold m ((a, Maybe a), (Ring a, Ptr a)) a
+sumWith = Fold step initial extract
+
+    where
+
+    initial =
+        return
+            $ Partial
+            $ Tuple'
+                (0 :: a) -- running sum
+                (0 :: a) -- accumulated rounding error
+
+    step (Tuple' total err) ((new0, mOld), _) =
+        let incr =
+                case mOld of
+                    -- XXX new may be large and err may be small we may lose it
+                    Nothing -> new0 - err
+                    -- XXX if (new - old) is large we may lose err
+                    Just old -> (new0 - old) - err
+            -- total is large and incr may be small, we may round incr here but
+            -- we will accumulate the rounding error in err1 in the next step.
+            total1 = total + incr
+            -- Accumulate any rounding error in err1
+            -- XXX In the Nothing case above we may lose err, therefore we
+            -- should use ((total1 - total) - new) + err here.
+            -- Or even in the just case if (new - old) is large we may lose
+            -- err, so we should use ((total1 - total) + (old - new)) + err.
+            err1 = (total1 - total) - incr
+        in return $ Partial $ Tuple' total1 err1
+
+    extract (Tuple' total _) = return total
+
+{-# INLINE lengthWith #-}
+lengthWith :: (Monad m, Num b)
+    => Fold m ((a, Maybe a), (Ring a, Ptr a)) b
+lengthWith = Fold.foldl' step 0
+
+    where
+
+    step w ((_, Nothing), _) = w + 1
+    step w _ = w
+
+{-# INLINE meanWith #-}
+meanWith :: forall m a. (Monad m, Fractional a)
+    => Fold m ((a, Maybe a), (Ring a, Ptr a)) a
+meanWith = Fold.teeWith (/) sumWith lengthWith
+
+{-# INLINE foldRingBuff #-}
+foldRingBuff :: (MonadIO m, Storable a)
+    => Fold m ((a, Maybe a), (Ring a, Ptr a)) (Ring a, Ptr a, Int)
+foldRingBuff = Fold step initial extract
+
+    where
+
+    initial = do
+        (rb0, rh0) <- liftIO $ new 0
+        return
+            $ Partial (rb0, rh0, 0)
+
+    step (_, _, i) ((_, _), (rb, rh)) =
+        return $ Partial (rb, rh, i+1)
+
+    extract = return
+
+md ::  MonadIO m => Int -> Fold m Double Double
+md n = slidingWindowWith n $ Fold.teeWith getMD meanWith foldRingBuff
```

Comment:
This is not good.

As I explained earlier we need to pass (a, Maybe a) to the `mean` fold and (Ring a, Ptr a) to the `last` fold and then zip the results. Use `lmap fst` and `lmap snd` to pass only selected input to `mean` and `last`. We do not need meanWith  or foldRingBuff, implementing such custom functions goes against all the modularity and makes it ugly.

Reply:
`md` should be implemented in the streamly-statistics package.

Reply:
how do we get the length of the ring buffer from last fold?
we need it to decide between using unsafeFoldRingM and unsafeFoldRingMFull.